### ISSUE TITLE
Segmentation Fault when setting x509 options in iothub_client_device_twin... sample

### ISSUE THREAD

**Issue by rschiefer** [Azure/azure-iot-sdk-c#2195]
Windows 10 - Ubuntu (20.04.1 LTS) in WSL
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

**SDK Version (Please Give Commit SHA if Manually Compiling)**

Tagged 1.4.0

**Protocol**

MQTT

**Describe the Bug**

SAS IoT Hub connection string works without issue but when I try to use a x509 connection and set the certificate options I get a segmentation error.

Code:

```
           (IoTHubDeviceClient_SetOption(iotHubClientHandle, OPTION_X509_CERT, x509certificate2) != IOTHUB_CLIENT_OK) ||
           (IoTHubDeviceClient_SetOption(iotHubClientHandle, OPTION_X509_PRIVATE_KEY , x509privatekey2) != IOTHUB_CLIENT_OK)

```
Valgrind Log:

```
==2133== Memcheck, a memory error detector
==2133== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2133== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==2133== Command: ./iothub_client_device_twin_and_methods_sample
==2133==
==2133== Conditional jump or move depends on uninitialised value(s)
==2133==    at 0x149673: IoTHubClientCore_SetOption (iothub_client_core.c:1705)
==2133==    by 0x142D25: IoTHubDeviceClient_SetOption (iothub_device_client.c:73)
==2133==    by 0x119A6F: iothub_client_device_twin_and_methods_sample_run (iothub_client_device_twin_and_methods_sample.c:421)
==2133==    by 0x119C1B: main (iothub_client_device_twin_and_methods_sample.c:486)
==2133==  Uninitialised value was created by a stack allocation
==2133==    at 0x119A13: iothub_client_device_twin_and_methods_sample_run (iothub_client_device_twin_and_methods_sample.c:391)
==2133==
==2133== Use of uninitialised value of size 8
==2133==    at 0x1496E0: IoTHubClientCore_SetOption (iothub_client_core.c:1719)
==2133==    by 0x142D25: IoTHubDeviceClient_SetOption (iothub_device_client.c:73)
==2133==    by 0x119A6F: iothub_client_device_twin_and_methods_sample_run (iothub_client_device_twin_and_methods_sample.c:421)
==2133==    by 0x119C1B: main (iothub_client_device_twin_and_methods_sample.c:486)
==2133==  Uninitialised value was created by a stack allocation
==2133==    at 0x119A13: iothub_client_device_twin_and_methods_sample_run (iothub_client_device_twin_and_methods_sample.c:391)
==2133==
==2133== Invalid read of size 8
==2133==    at 0x1496E0: IoTHubClientCore_SetOption (iothub_client_core.c:1719)
==2133==    by 0x142D25: IoTHubDeviceClient_SetOption (iothub_device_client.c:73)
==2133==    by 0x119A6F: iothub_client_device_twin_and_methods_sample_run (iothub_client_device_twin_and_methods_sample.c:421)
==2133==    by 0x119C1B: main (iothub_client_device_twin_and_methods_sample.c:486)
==2133==  Address 0x111a is not stack'd, malloc'd or (recently) free'd
==2133==
==2133==
==2133== Process terminating with default action of signal 11 (SIGSEGV)
==2133==  Access not within mapped region at address 0x111A
==2133==    at 0x1496E0: IoTHubClientCore_SetOption (iothub_client_core.c:1719)
==2133==    by 0x142D25: IoTHubDeviceClient_SetOption (iothub_device_client.c:73)
==2133==    by 0x119A6F: iothub_client_device_twin_and_methods_sample_run (iothub_client_device_twin_and_methods_sample.c:421)
==2133==    by 0x119C1B: main (iothub_client_device_twin_and_methods_sample.c:486)
==2133==  If you believe this happened as a result of a stack
==2133==  overflow in your program's main thread (unlikely but
==2133==  possible), you can try to increase the size of the
==2133==  main thread stack using the --main-stacksize= flag.
==2133==  The main thread stack size used in this run was 8388608.
==2133==
==2133== HEAP SUMMARY:
==2133==     in use at exit: 202,194 bytes in 4,369 blocks
==2133==   total heap usage: 6,438 allocs, 2,069 frees, 280,609 bytes allocated
==2133==
==2133== LEAK SUMMARY:
==2133==    definitely lost: 0 bytes in 0 blocks
==2133==    indirectly lost: 0 bytes in 0 blocks
==2133==      possibly lost: 0 bytes in 0 blocks
==2133==    still reachable: 202,194 bytes in 4,369 blocks
==2133==         suppressed: 0 bytes in 0 blocks
==2133== Reachable blocks (those to which a pointer was found) are not shown.
==2133== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==2133==
==2133== For lists of detected and suppressed errors, rerun with: -s
==2133== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)
Segmentation fault
```

**Comment by danewalton** [comment]
Hi @rschiefer 
It looks like line numbers are being output from the sample which don't exist in the default sample. I'm assuming you've added your own logic to the sample?

With that, I'm guessing somehow a new variable is being created and isn't being allocated correctly to then be passed to the SDK. Are you able to share the sample as a whole with your personal info redacted?

**Comment by rschiefer** [comment]
Yes, I'm creating 2 variables right below the connectionstring variable (line 58)  for the x509 cert and private key:

```

static const char* x509certificate2 =
"-----BEGIN CERTIFICATE-----""\n"
"MIIFkDCCA3igAwIBAgIBAzANBgkqhkiG9w0BAQsFADA0MTIwMAYDVQQDDClBenVy""\n"
"ZSBJb1QgSHViIEludGVybWVkaWF0ZSBDZXJ0IFRlc3QgT25seTAeFw0yMjAxMDMy""\n"
"MDI2MjNaFw0yMzAxMDMyMDI2MjNaMBsxGTAXBgNVBAMMEHNpbXVsYXRlZERldmlj""\n"...
```

is this right?

**Comment by danewalton** [comment]
Can you share the larger code sample that you're running? It's hard to parse out what variable might be causing this.

**Comment by anhashia** [comment]
We will look at it. Thanks !

**Comment by danewalton** [comment]
Hi @rschiefer 
I modified the `iothub_convenience_sample.c` to use cert authentication and it looks to be working for me. [Linked here is the diff of the code which I ran. Please have a look and hopefully it can help you parse where things might be going wrong.](https://github.com/Azure/azure-iot-sdk-c/compare/gh2195)

**Comment by danewalton** [comment]
Note I tried this same thing with `iothub_client_device_twin_and_methods` and it works as well.

**Comment by rschiefer** [comment]
Here is the a gist of the code I'm using that results in a seg fault.  

https://gist.github.com/rschiefer/1277730b3083ee3d4ebd3c173d42421d

I'm new to C so I'm probably doing something wrong but don't see it.

**Comment by rschiefer** [comment]
I got it to work by moving the setoptions for cert under the createConnection call.  Sorry for my confusion.  Thanks for the help!

**Comment by danewalton** [comment]
Glad we could help!
